scrapy_zyte_api/_downloader_middleware.py

```diff
@@ -23,3 +36,22 @@ def process_request(self, request, spider):
             request.meta["download_slot"] = slot_id
         _, slot = downloader._get_slot(request, spider)
         slot.delay = 0
+
+        if self._max_requests_reached(downloader):
+            self._crawler.engine.close_spider(spider, "closespider_max_zapi_requests")
+            raise IgnoreRequest("Reached max Zyte API requests")
+
+    def _max_requests_reached(self, downloader) -> bool:
+        if not self._max_requests:
+            return False
+
+        zapi_req_count = self._crawler.stats.get_value("scrapy-zyte-api/processed", 0)
+        download_req_count = sum(
+            [
+                len(slot.transferring)
```

Comment:
This is a tricky one since if one of the ongoing requests fail (after some best efforts to have them retried), then the _successful billable ZAPI requests_ **<** `ZYTE_API_MAX_REQUESTS`.

Perhaps this would be okay?

Reply:
Do you mean "the user wanted 1000 requests but only used 990"? I think that's fine.

Reply:
where they wanted exactly 1000 billable ZAPI requests but only 990 were billed as 10 had errors in them.

Reply:
This behavior sounds fine to me.